### PR TITLE
AE-1250 Comp Delegations Monitor

### PR DESCRIPTION
**Pull request by darian-ae**
This PR creates a new bot that monitors for when a given address has been delegated enough COMP to pass critical governance thresholds. It is similar to the `large-borrows-governance` bot in the way that it's written and uses those same governance threshold numbers (I tried to squeeze it in there at first but it got messy fast and introduced some bugs so I created a separate bot here)

### CODE REVIEW

File: comp-delegations-monitor/.dockerignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+node_modules
+dist
+forta.config.json
```
Comment (Nomad-AE) [inline comp-delegations-monitor/.dockerignore:3]:
```suggestion
forta.config.json

```
minor nit.

File: comp-delegations-monitor/.gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+node_modules
+dist
+forta.config.json
```
Comment (Nomad-AE) [inline comp-delegations-monitor/.gitignore:3]:
```suggestion
forta.config.json

```
nit

File: comp-delegations-monitor/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+WORKDIR /app
+# if using obfuscated code from build stage:
+# COPY --from=builder /app/dist ./src
+# else if using unobfuscated code:
+COPY ./abi ./abi
+COPY ./src ./src
+COPY agent-config.json ./
+COPY package*.json ./
+RUN npm ci --production
+CMD [ "npm", "run", "start:prod" ]
```
Comment (Nomad-AE) [inline comp-delegations-monitor/Dockerfile:20]:
```suggestion
CMD [ "npm", "run", "start:prod" ]

```
nit

File: comp-delegations-monitor/src/agent.spec.js [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+  it('has a valid governor bravo address and api ', () => {
+    const { governorBravo } = config;
+    const { address, abi } = governorBravo;
+    expect(ethers.utils.isHexString(address, 20)).toBe(true);
+    expect(typeof (abi)).toBe('string');
+    expect(abi).not.toBe('');
+  });
+
+  describe('monitor COMP token delegation thresholds', () => {
+    let inittializeData;
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:169]:
```suggestion
    let initializeData;
```
typo

File: comp-delegations-monitor/src/agent.spec.js [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    let handleTransaction;
+    let mockTxEvent;
+
+    const compERC20Address = '0xc00e94Cb662C3520282E6f5717214004A7f26888';
+    const mockDelegateAddress = '0x1212121212121212121212121212121212121212';
+    const delegateEventName = 'DelegateVotesChanged';
+    const nonDelegateEventName = 'Transfer';
+
+    beforeEach(async () => {
+      inittializeData = {};
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:179]:
```suggestion
      initializeData = {};
```
typo

File: comp-delegations-monitor/src/agent.spec.js [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+      // initialize handler
+      await (provideInitialize(inittializeData))();
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:182]:
```suggestion
      await (provideInitialize(initializeData))();
```
typo

File: comp-delegations-monitor/src/agent.spec.js [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+      handleTransaction = provideHandleTransaction(inittializeData);
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:183]:
```suggestion
      handleTransaction = provideHandleTransaction(initializeData);
```
typo

File: comp-delegations-monitor/src/agent.spec.js [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+        { address: zeroAddress },
+      );
+
+      mockTxEvent.logs = [log];
+
+      const findings = await handleTransaction(mockTxEvent);
+      expect(findings).toStrictEqual([]);
+    });
+
+    it('returns no findings if no DelegateVotesChanged event was emitted, but transaction involed the COMP token address', async () => {
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:205]:
```suggestion
    it('returns no findings if no DelegateVotesChanged event was emitted, but transaction involved the COMP token address', async () => {
```
typo

File: comp-delegations-monitor/src/agent.spec.js [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+        { address: compERC20Address },
+      );
+
+      mockTxEvent.logs = [log];
+
+      const findings = await handleTransaction(mockTxEvent);
+      expect(findings).toStrictEqual([]);
+    });
+
+    it('returns no findings if a DelegateVotesChanged event occured with the COMP token, but threshold level was not crossed', async () => {
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:221]:
```suggestion
    it('returns no findings if a DelegateVotesChanged event occurred with the COMP token, but threshold level was not crossed', async () => {
```
typo

File: comp-delegations-monitor/src/agent.spec.js [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+      mockGovContract.balanceOf.mockResolvedValue(currCOMPOwned.multipliedBy(decimals));
+
+      const log = createLog(
+        compERC20Interface.getEvent(delegateEventName),
+        {
+          delegate: mockDelegateAddress,
+          previousBalance: 0,
+          newBalance: currCOMPOwned.multipliedBy(decimals).toString(),
+        },
+        { address: compERC20Address }, // input invloved address for .filterLog()
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:234]:
```suggestion
        { address: compERC20Address }, // input involved address for .filterLog()
```
typo

File: comp-delegations-monitor/src/agent.spec.js [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+      );
+
+      mockTxEvent.logs = [log];
+
+      const finding = await handleTransaction(mockTxEvent);
+      expect(finding).toStrictEqual([]);
+    });
+
+    it('returns a proposal threshold finding if delegate balance crosses proposal threshold, but not quorum thresdhold', async () => {
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:243]:
```suggestion
    it('returns a proposal threshold finding if delegate balance crosses proposal threshold, but not quorum threshold', async () => {
```
typo

File: comp-delegations-monitor/src/agent.spec.js [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+      mockGovContract.balanceOf.mockResolvedValue(currCOMPOwned.multipliedBy(decimals));
+
+      const log = createLog(
+        compERC20Interface.getEvent(delegateEventName),
+        {
+          delegate: mockDelegateAddress,
+          previousBalance: 0,
+          newBalance: currCOMPOwned.multipliedBy(decimals).toString(),
+        },
+        { address: compERC20Address }, // input invloved address for .filterLog()
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:256]:
```suggestion
        { address: compERC20Address }, // input involved address for .filterLog()
```

File: comp-delegations-monitor/src/agent.spec.js [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
```
Comment (Nomad-AE) [inline comp-delegations-monitor/src/agent.spec.js:296]:
```suggestion
        { address: compERC20Address }, // input involved address for .filterLog()
```
typo

**Review by Nomad-AE — APPROVED** [review]
Code looks good, just a few spelling errors and small nit-picks.

File: comp-delegations-monitor/src/agent.spec.js [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+  });
+
+  // assign topic and data
+  log.topics = topics;
+  log.data = ethers.utils.defaultAbiCoder.encode(dataTypes, dataValues);
+
+  return log;
+}
+
+describe('check agent configuration file', () => {
```
Comment (arbitrary-sparrow) [inline comp-delegations-monitor/src/agent.spec.js:133]:
```suggestion
describe('check bot configuration file', () => {
```

File: comp-delegations-monitor/agent-config.json [hunk 15]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (arbitrary-sparrow) [inline comp-delegations-monitor/agent-config.json:1]: This file should be renamed to `bot-config.json`

File: comp-delegations-monitor/src/agent.js [hunk 16]
```diff
@@ -0,0 +1,159 @@
+/* eslint-disable global-require */
+const {
+  ethers, Finding, FindingSeverity, FindingType, getEthersProvider,
+} = require('forta-agent');
+
+const BigNumber = require('bignumber.js');
+
+const config = require('../agent-config.json');
```
Comment (arbitrary-sparrow) [inline comp-delegations-monitor/src/agent.js:8]:
```suggestion
const config = require('../bot-config.json');
```

**Review by arbitrary-sparrow — CHANGES_REQUESTED** [review]
The overall logic looks fine, I just had some minor changes.

File: comp-delegations-monitor/.dockerignore (hunk 1 above)
Comment (darian-ae) [reply comp-delegations-monitor/.dockerignore:3]: I did not know that, thanks for explaining!

**Review by arbitrary-sparrow — APPROVED** [review]

**Pull request merged**